### PR TITLE
Add parameter validation to pk module

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Add null-pointer-parameter validation for the PK module (including pkparse and pkwrite).

### CODE REVIEW

File: library/pk.c [hunk 1]
```diff
@@ -175,7 +184,6 @@ int mbedtls_pk_setup_rsa_alt( mbedtls_pk_context *ctx, void * key,
  */
 int mbedtls_pk_can_do( const mbedtls_pk_context *ctx, mbedtls_pk_type_t type )
 {
-    /* null or NONE context can't do anything */
```
Comment (Patater) [inline library/pk.c]: Why remove this comment (in "Implement parameter validation in pk, pkparse and pkwrite")? It helps to explain why `PK_VALIDATE` is not being used here.

File: library/pkparse.c [hunk 2]
```diff
@@ -74,6 +80,10 @@ int mbedtls_pk_load_file( const char *path, unsigned char **buf, size_t *n )
     FILE *f;
     long size;
 
+    PK_VALIDATE_RET( path != NULL );
+    PK_VALIDATE_RET( buf != NULL );
```
Comment (Patater) [inline library/pkparse.c]: Would additionally checking `*buf` be within the current design philosophy of the check params feature?

File: library/pkparse.c [hunk 3]
```diff
@@ -1145,12 +1164,17 @@ int mbedtls_pk_parse_key( mbedtls_pk_context *pk,
...
-
 #if defined(MBEDTLS_PEM_PARSE_C)
     size_t len;
     mbedtls_pem_context pem;
+#endif
 
-    mbedtls_pem_init( &pem );
+    PK_VALIDATE_RET( pk != NULL );
+    PK_VALIDATE_RET( key != NULL || keylen == 0 );
+    PK_VALIDATE_RET( pwd != NULL || pwdlen == 0 );
```
Comment (Patater) [inline library/pkparse.c]: Do we test that this function supports `keylen == 0` with a null `key` (same for null `pwd` with `pwdlen == 0`)? It looks like at least most paths through this function, depending on the configuration, will also check `keylen == 0` and reject it. If we already have checks for this elsewhere in the function, why add another check here?

File: library/pkparse.c [hunk 4]
```diff
@@ -1360,7 +1384,12 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
 #if defined(MBEDTLS_PEM_PARSE_C)
     size_t len;
     mbedtls_pem_context pem;
+#endif
+
+    PK_VALIDATE_RET( ctx != NULL );
+    PK_VALIDATE_RET( key != NULL || keylen == 0 );
```
Comment (Patater) [inline library/pkparse.c]: Depending on the library configuration, this function checks for `keylen == 0` later on and rejects it. Do we need another check here? Could we remove the latter check?

File: include/mbedtls/pk.h [hunk 5]
```diff
@@ -197,23 +197,35 @@ typedef size_t (*mbedtls_pk_rsa_alt_key_len_func)( void *ctx );
...
 void mbedtls_pk_restart_init( mbedtls_pk_restart_ctx *ctx );
 
 /**
  * \brief           Free the components of a restart context
+ *
+ * \param ctx       The context to clear.
+ *                  If this is \c NULL, this function does nothing.
  */
 void mbedtls_pk_restart_free( mbedtls_pk_restart_ctx *ctx );
 #endif /* MBEDTLS_ECDSA_C && MBEDTLS_ECP_RESTARTABLE */
```
Comment (hanno-arm) [inline include/mbedtls/pk.h]: All the other public functions in the PK API need to be properly documented, too.

File: library/pk.c [hunk 6]
```diff
@@ -151,7 +159,8 @@ int mbedtls_pk_setup_rsa_alt( mbedtls_pk_context *ctx, void * key,
     mbedtls_rsa_alt_context *rsa_alt;
     const mbedtls_pk_info_t *info = &mbedtls_rsa_alt_info;
 
-    if( ctx == NULL || ctx->pk_info != NULL )
+    PK_VALIDATE_RET( ctx != NULL );
```
Comment (hanno-arm) [inline library/pk.c]: Does it make sense to use RSA ALT contexts with NULL key, e.g. when wrapping an HSM? In this case, I'm OK with not adding more checks, but the documentation should definitely mention the requirements here.

File: library/pk.c [hunk 7]
```diff
@@ -175,7 +184,6 @@ int mbedtls_pk_setup_rsa_alt( mbedtls_pk_context *ctx, void * key,
  */
 int mbedtls_pk_can_do( const mbedtls_pk_context *ctx, mbedtls_pk_type_t type )
 {
-    /* null or NONE context can't do anything */
     if( ctx == NULL || ctx->pk_info == NULL )
```
Comment (hanno-arm) [inline library/pk.c]: This check should be moved to parameter validation.

File: library/pk.c [hunk 8]
```diff
@@ -232,7 +240,11 @@ int mbedtls_pk_verify_restartable( mbedtls_pk_context *ctx,
                const unsigned char *sig, size_t sig_len,
```
Comment (hanno-arm) [inline library/pk.c]: Non-blocker: Consider fixing the indentation while at it.

File: library/pk.c [hunk 9]
```diff
@@ -232,7 +240,11 @@ int mbedtls_pk_verify_restartable( mbedtls_pk_context *ctx,
...
                mbedtls_pk_restart_ctx *rs_ctx )
 {
-    if( ctx == NULL || ctx->pk_info == NULL ||
+    PK_VALIDATE_RET( ctx != NULL );
+    PK_VALIDATE_RET( hash != NULL );
+    PK_VALIDATE_RET( sig != NULL );
```
Comment (hanno-arm) [inline library/pk.c]: We should require `( md_alg == MBEDTLS_MD_NONE && hash_len == 0 ) || hash != NULL`.

File: library/pk.c [hunk 10]
```diff
@@ -285,7 +297,11 @@ int mbedtls_pk_verify_ext( mbedtls_pk_type_t type, const void *options,
                    const unsigned char *hash, size_t hash_len,
                    const unsigned char *sig, size_t sig_len )
 {
-    if( ctx == NULL || ctx->pk_info == NULL )
+    PK_VALIDATE_RET( ctx != NULL );
+    PK_VALIDATE_RET( hash != NULL );
```
Comment (hanno-arm) [inline library/pk.c]: Same here.

File: library/pk.c [hunk 11]
```diff
@@ -285,7 +297,11 @@ int mbedtls_pk_verify_ext( mbedtls_pk_type_t type, const void *options,
...
+    PK_VALIDATE_RET( sig != NULL );
+
+    if( ctx->pk_info == NULL )
```
Comment (hanno-arm) [inline library/pk.c]: In line with the cipher and md modules, I think this should be part of parameter validation, too.

File: library/pk.c [hunk 12]
```diff
@@ -345,7 +361,11 @@ int mbedtls_pk_sign_restartable( mbedtls_pk_context *ctx,
              int (*f_rng)(void *, unsigned char *, size_t), void *p_rng,
              mbedtls_pk_restart_ctx *rs_ctx )
 {
-    if( ctx == NULL || ctx->pk_info == NULL ||
+    PK_VALIDATE_RET( ctx != NULL );
+    PK_VALIDATE_RET( hash != NULL );
+    PK_VALIDATE_RET( sig != NULL );
+
+    if( ctx->pk_info == NULL ||
```
Comment (hanno-arm) [inline library/pk.c]: I think this should be part of parameter validation, too.

File: library/pk.c [hunk 13]
```diff
@@ -345,7 +361,11 @@ int mbedtls_pk_sign_restartable( mbedtls_pk_context *ctx,
              int (*f_rng)(void *, unsigned char *, size_t), void *p_rng,
              mbedtls_pk_restart_ctx *rs_ctx )
 {
-    if( ctx == NULL || ctx->pk_info == NULL ||
+    PK_VALIDATE_RET( ctx != NULL );
+    PK_VALIDATE_RET( hash != NULL );
```
Comment (hanno-arm) [inline library/pk.c]: As above.

File: library/pk.c [hunk 14]
```diff
@@ -399,7 +419,12 @@ int mbedtls_pk_decrypt( mbedtls_pk_context *ctx,
                 unsigned char *output, size_t *olen, size_t osize,
                 int (*f_rng)(void *, unsigned char *, size_t), void *p_rng )
 {
-    if( ctx == NULL || ctx->pk_info == NULL )
+    PK_VALIDATE_RET( ctx != NULL );
+    PK_VALIDATE_RET( input != NULL || ilen == 0 );
+    PK_VALIDATE_RET( output != NULL || osize == 0 );
+    PK_VALIDATE_RET( olen != NULL );
+
+    if( ctx->pk_info == NULL )
```
Comment (hanno-arm) [inline library/pk.c]: I think this should be part of parameter validation.

File: library/pk.c [hunk 15]
```diff
@@ -417,7 +442,12 @@ int mbedtls_pk_encrypt( mbedtls_pk_context *ctx,
...
```
Comment (hanno-arm) [inline library/pk.c]: As above.

File: include/mbedtls/pk.h (hunk 5 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/pk.h]: What's the objective for “properly documented”? I could spend a long time improving the documentation, but we're pressed for time here, so I'm trying to stick to the necessary minimum.

Comment (hanno-arm) [reply include/mbedtls/pk.h]: Concretely, I suggest to apply the same kind of documentation enhancements as in the other parameter validation PRs.

The situation is far from ideal, for example because it doesn't distinguish between conditions we check and such we just assume, but it's still an improvement over the existing documentation.

File: library/pkparse.c (hunk 2 above)
Comment (hanno-arm) [reply library/pkparse.c]: No, because `*buf` is being set by this function; its initial state doesn't matter.

File: library/pk.c [hunk 16]
```diff
@@ -128,7 +135,8 @@ const mbedtls_pk_info_t * mbedtls_pk_info_from_type( mbedtls_pk_type_t pk_type )
  */
 int mbedtls_pk_setup( mbedtls_pk_context *ctx, const mbedtls_pk_info_t *info )
 {
-    if( ctx == NULL || info == NULL || ctx->pk_info != NULL )
+    PK_VALIDATE_RET( ctx != NULL );
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply library/pk.c]: Discussed internally and dismissed.

File: library/pk.c (hunk 11 above)
Comment (hanno-arm) [reply library/pk.c]: Discussed internally and dismissed. The MD and Cipher module PRs will need to be reworked.

File: library/pk.c (hunk 12 above)
Comment (hanno-arm) [reply library/pk.c]: Discussed internally and dismissed. The MD and Cipher module PRs will need to be reworked.

File: library/pk.c (hunk 15 above)
Comment (hanno-arm) [reply library/pk.c]: Discussed internally and dismissed. The MD and Cipher module PRs will need to be reworked.

File: library/pkparse.c [hunk 17]
```diff
@@ -124,6 +134,8 @@ int mbedtls_pk_parse_keyfile( mbedtls_pk_context *ctx,
     size_t n;
     unsigned char *buf;
 
+    PK_VALIDATE_RET( ctx != NULL );
```
Comment (hanno-arm) [inline library/pkparse.c]: We should check `path != NULL`, too.

File: library/pkparse.c [hunk 18]
```diff
@@ -148,6 +160,8 @@ int mbedtls_pk_parse_public_keyfile( mbedtls_pk_context *ctx, const char *path )
...
```
Comment (hanno-arm) [inline library/pkparse.c]: We should check for `path != NULL` here (even if we check it in `mbedtls_pk_load_file()`, too - that's an implementation detail).

File: library/pkparse.c (hunk 3 above)
Comment (hanno-arm) [reply library/pkparse.c]: You might dismiss this as theoretical babbling, but I think the question of which checks we perform as part of parameter validation should solely depend on the signature and semantics of the respective function, and not on how it's implemented. Also, if you care about code-size, you don't enable this feature in the first place.

File: tests/suites/test_suite_pk.function [hunk 19]
```diff
@@ -69,6 +69,322 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
...
 
+/* BEGIN_CASE */
+void valid_parameters( )
+{
+    mbedtls_pk_context pk;
+    size_t len;
+
+    mbedtls_pk_init( &pk );
+
+    TEST_VALID_PARAM( mbedtls_pk_setup( &pk, NULL ) ==
```
Comment (hanno-arm) [inline tests/suites/test_suite_pk.function]: Please add a test for `mbedtls_pk_free( NULL )`.

File: tests/suites/test_suite_pk.function [hunk 20]
```diff
@@ -69,6 +69,322 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
...
+                      MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+
+    TEST_VALID_PARAM( mbedtls_pk_get_bitlen( NULL ) == 0 );
+
+    TEST_VALID_PARAM( mbedtls_pk_get_len( NULL ) == 0 );
+
+    TEST_VALID_PARAM( mbedtls_pk_can_do( NULL, MBEDTLS_PK_NONE ) == 0 );
```
Comment (hanno-arm) [inline tests/suites/test_suite_pk.function]: As mentioned, I think it should be forbidden to pass `NULL` to `mbedtls_pk_can_do()`, as for the rest of the PK API. In particular, this test should be moved to `invalid_parameters()` as a negative test.

File: library/pk.c [hunk 21]
```diff
@@ -84,7 +90,8 @@ void mbedtls_pk_restart_init( mbedtls_pk_restart_ctx *ctx )
  */
 void mbedtls_pk_restart_free( mbedtls_pk_restart_ctx *ctx )
 {
-    if( ctx == NULL || ctx->pk_info == NULL ||
+    PK_VALIDATE( ctx != NULL );
```
Comment (hanno-arm) [inline library/pk.c]: It can be argued whether it's right, but we have decided to document that `xxx_free()` functions accept `NULL` parameters and behave as no-ops in this case.

File: tests/suites/test_suite_pk.function [hunk 22]
```diff
@@ -69,6 +69,322 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
...
+    void *options = buf;
+
+    mbedtls_pk_init( &pk );
+
+    TEST_INVALID_PARAM( mbedtls_pk_init( NULL ) );
+
+#if defined(MBEDTLS_ECDSA_C) && defined(MBEDTLS_ECP_RESTARTABLE)
+    TEST_INVALID_PARAM( mbedtls_pk_restart_init( NULL ) );
+
+    TEST_INVALID_PARAM( mbedtls_pk_restart_free( NULL ) );
```
Comment (hanno-arm) [inline tests/suites/test_suite_pk.function]: This should be a positive test in `valid_parameters()`.

File: tests/suites/test_suite_pk.function [hunk 23]
```diff
@@ -69,6 +69,322 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
...
+{
+    mbedtls_pk_context pk;
+    size_t len;
+
+    mbedtls_pk_init( &pk );
+
+    TEST_VALID_PARAM( mbedtls_pk_setup( &pk, NULL ) ==
+                      MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+
+    TEST_VALID_PARAM( mbedtls_pk_get_bitlen( NULL ) == 0 );
```
Comment (hanno-arm) [inline tests/suites/test_suite_pk.function]: As mentioned, I think it should be forbidden to pass `NULL` to `mbedtls_pk_can_do()`, as for the rest of the PK API. While parameter validation isn't suited to test this because it doesn't return `int`, and so we might not want to change the code for `mbedtls_pk_get_bitlen()`, at least we shouldn't add a test here suggesting that passing `NULL` is ok.

File: tests/suites/test_suite_pk.function [hunk 24]
```diff
@@ -69,6 +69,322 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
...
+
+    TEST_VALID_PARAM( mbedtls_pk_get_len( NULL ) == 0 );
```
Comment (hanno-arm) [inline tests/suites/test_suite_pk.function]: Same here. Please remove this test.

File: library/pkparse.c (hunk 3 above)
Comment (Patater) [reply library/pkparse.c]: How its implemented matters because we may break how people expect it to work. We can say "sorry, we didn't promise that in the docs", but we'd still potentially be breaking applications. See https://news.ycombinator.com/item?id=2281932 for a fun story.

Comment (hanno-arm) [reply library/pkparse.c]: @Patater I replied to your suggestion of removing checks if they are already performed deeper in the call stack, which I think shouldn't be done.

File: library/pkparse.c (hunk 2 above)
Comment (Patater) [reply library/pkparse.c]: Yup, agreed.

File: library/pk.c (hunk 1 above)
Comment (gilles-peskine-arm) [reply library/pk.c]: I found that comment unclear: what's “null or NONE” in this context? And once I understood that it meant “NULL or with type MBEDTLS_PK_NONE”, I found it uninformative: yes, I can see that it can't do anything, that's what it means to have can_do return false. I've now added a comment that explains _why_ can_do behaves this way.

File: library/pkparse.c (hunk 3 above)
Comment (gilles-peskine-arm) [reply library/pkparse.c]: It is tested in the `valid_parameters` test.

File: library/pkparse.c (hunk 4 above)
Comment (gilles-peskine-arm) [reply library/pkparse.c]: I've refactored the `keylen==0` checks to make the behavior more obvious: reject `keylen==0` early and unconditionally instead of testing it multiple times later.

File: include/mbedtls/pk.h [hunk 25]
```diff
@@ -197,23 +197,35 @@ typedef size_t (*mbedtls_pk_rsa_alt_key_len_func)( void *ctx );
...
+ *                  This must not be \c NULL.
  */
 void mbedtls_pk_init( mbedtls_pk_context *ctx );
 
 /**
- * \brief           Free a mbedtls_pk_context
+ * \brief           Free the components of a #mbedtls_pk_context.
+ *
+ * \param ctx       The context to clear.
+ *                  If this is \c NULL, this function does nothing.
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/mbedtls/pk.h]: Added

File: include/mbedtls/pk.h [hunk 26]
```diff
@@ -197,23 +197,35 @@ typedef size_t (*mbedtls_pk_rsa_alt_key_len_func)( void *ctx );
...
+ * \param ctx       The context to initialize.
+ *                  This must not be \c NULL.
  */
 void mbedtls_pk_restart_init( mbedtls_pk_restart_ctx *ctx );
 
 /**
  * \brief           Free the components of a restart context
+ *
+ * \param ctx       The context to clear.
+ *                  If this is \c NULL, this function does nothing.
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/mbedtls/pk.h]: Added

File: include/mbedtls/pk.h (hunk 5 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/pk.h]: I've added initialized/setup/private requirements for the context argument in each function.

File: library/pk.c (hunk 6 above)
Comment (gilles-peskine-arm) [reply library/pk.c]: I don't know, but in doubt, I prefer to not change the behavior.

File: library/pk.c (hunk 7 above)
Comment (gilles-peskine-arm) [reply library/pk.c]: `can_do` is a function that one might plausibly call early on before bailing out because allocating the context failed, and returning 0 here is safe, so I prefer to err on the side of backward compatibility. We can officially deprecate it later and change the behavior after that.

File: library/pk.c (hunk 8 above)
Comment (gilles-peskine-arm) [reply library/pk.c]: I prefer not to change unrelated things. And by the way as a reviewer I prefer when unrelated things don't change.

File: library/pk.c (hunk 9 above)
Comment (gilles-peskine-arm) [reply library/pk.c]: Good point. Fixed and I added a corresponding test.

File: library/pk.c (hunk 10 above)
Comment (gilles-peskine-arm) [reply library/pk.c]: Also fixed and tested

File: library/pk.c (hunk 13 above)
Comment (gilles-peskine-arm) [reply library/pk.c]: also fixed and tested

File: library/pkparse.c (hunk 17 above)
Comment (gilles-peskine-arm) [reply library/pkparse.c]: It was done from `load_file`, but ok, it's clearer that way.

File: library/pkparse.c (hunk 18 above)
Comment (gilles-peskine-arm) [reply library/pkparse.c]: as above

File: tests/suites/test_suite_pk.function (hunk 19 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function]: done

File: library/pk.c (hunk 7 above)
Comment (hanno-arm) [reply library/pk.c]: Fine by me.

File: library/pk.c (hunk 21 above)
Comment (gilles-peskine-arm) [reply library/pk.c]: My mistake. Fixed.

File: library/pk.c (hunk 6 above)
Comment (hanno-arm) [reply library/pk.c]: Ok.

File: tests/suites/test_suite_pk.function (hunk 22 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function]: fixed

File: tests/suites/test_suite_pk.function (hunk 20 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function]: I'm keeping the test: this is undocumented behavior, but behavior that we shouldn't change without meaning to. I'll add a comment to this effect.

File: tests/suites/test_suite_pk.function (hunk 24 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function]: Same as `can_do`

File: tests/suites/test_suite_pk.function (hunk 23 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function]: Same as `can_do`

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function]: Same as `can_do`

File: tests/suites/test_suite_pk.function [hunk 27]
```diff
@@ -432,10 +432,6 @@ void invalid_parameters( )
                             mbedtls_pk_parse_key( &pk,
                                                   NULL, sizeof( buf ),
                                                   buf, sizeof( buf ) ) );
-    TEST_INVALID_PARAM_RET( MBEDTLS_ERR_PK_BAD_INPUT_DATA,
-                            mbedtls_pk_parse_key( &pk,
-                                                  buf, sizeof( buf ),
-                                                  NULL, sizeof( buf ) ) );
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function]: Yes:
```
    TEST_ASSERT( mbedtls_pk_parse_key( &pk, NULL, 0, NULL, 1 ) ==
                 MBEDTLS_ERR_PK_KEY_INVALID_FORMAT );
```
I removed it precisely because it is valid and that test failed.

File: tests/suites/test_suite_pk.function [hunk 28]
```diff
@@ -155,6 +200,7 @@ void invalid_parameters( )
     unsigned char *p = buf;
     char str[1] = {0};
     mbedtls_pk_context pk;
+    mbedtls_md_type_t valid_md = MBEDTLS_MD_SHA256;
```
[the remark this answers is not in the document]
Comment (Patater) [reply tests/suites/test_suite_pk.function]: OK

File: tests/suites/test_suite_pk.function (hunk 27 above)
[the remark this answers is not in the document]
Comment (Patater) [reply tests/suites/test_suite_pk.function]: OK

File: library/pkwrite.c [hunk 29]
```diff
@@ -457,6 +481,9 @@ int mbedtls_pk_write_pubkey_pem( mbedtls_pk_context *key, unsigned char *buf, si
     unsigned char output_buf[PUB_DER_MAX_BYTES];
     size_t olen = 0;
 
+    PK_VALIDATE_RET( key != NULL );
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/pkwrite.c]: I chose to let the PEM function do their thing (null-check the parameters) and let the `pem_write` function worry about the size.

File: library/pkparse.c [hunk 30]
```diff
@@ -1360,11 +1384,18 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
 #if defined(MBEDTLS_PEM_PARSE_C)
     size_t len;
     mbedtls_pem_context pem;
+#endif
+
+    PK_VALIDATE_RET( ctx != NULL );
+    if( keylen == 0 )
+        return( MBEDTLS_ERR_PK_KEY_INVALID_FORMAT );
+    PK_VALIDATE_RET( key != NULL || keylen == 0 );
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/pkparse.c]: Oops, I forgot to change it. That's not a blocker so I won't rock the boat.

**Comment by sbutcher-arm** [comment]
Reviewed, approved and passing CI, to labelling as 'ready to merge'.

File: tests/suites/test_suite_pk.function [hunk 31]
```diff
@@ -200,7 +200,9 @@ void invalid_parameters( )
     unsigned char *null_buf = NULL;
```
Comment (hanno-arm) [inline tests/suites/test_suite_pk.function]: This is only needed if `PK_PARSE_C` is defined and will lead to an unused variable warning otherwise. I'd suggest to add `((void) xxx)` statements throughout to guard against such issues.

Comment (hanno-arm) [reply tests/suites/test_suite_pk.function]: The same holds for `p` as far as I see.

**Comment by sbutcher-arm** [comment]
Jenkins CI has two failures related to flakey DTLS tests, which can be accepted as a pass.

The commits are approved, therefore can keep the 'ready to merge' label.
